src/jdk.incubator.concurrent/share/classes/jdk/incubator/concurrent/StructuredTaskScope.java

```diff
@@ -0,0 +1,1175 @@
+ * a method to retrieve the exception of the first task to fail.
+ *
+ * <p> The following is an example of a StructuredTaskScope implementation that collects the
+ * results of tasks that complete successfully. It defines the method <b>{@code results()}</b>
+ * to be used by the main task to retrieve the results.
+ *
+ * {@snippet lang=java :
+ *     class MyScope<T> extends StructuredTaskScope<T> {
+ *         private final Queue<T> results = new ConcurrentLinkedQueue<>();
+ *
+ *         MyScope() {
+ *             super(null, Thread.ofVirtual().factory());
+ *         }
+ *
+ *         @Override
+ *         // @link substring="handleComplete" target="handleComplete" :
+ *         protected void handleComplete(Future<T> future) {
+ *             if (future.state() == Future.State.SUCCESS) {
+ *                 T result = future.resultNow();
+ *                 results.add(result);
+ *             }
+ *         }
+ *
+ *         // Returns a stream of results from tasks that completed successfully.
+ *         public Stream<T> results() {     // @highlight substring="results"
+ *             return results.stream();
+ *         }
+ *     }
+ *  }
+ *
+ * <h2><a id="TreeStructure">Tree structure</a></h2>
+ *
+ * StructuredTaskScopes form a tree where parent-child relations are established
+ * implicitly when opening a new task scope:
+ * <ul>
+ *   <li> A parent-child relation is established when a thread started in a task scope
+ *   opens its own task scope. A thread started in task scope "A" that opens task scope
+ *   "B" establishes a parent-child relation where task scope "A" is the parent of task
+ *   scope "B".
+ *   <li> A parent-child relation is established with nesting. If a thread opens task
+ *   scope "B", then opens task scope "C" (before it closes "B"), then the enclosing task
+ *   scope "B" is the parent of the nested task scope "C".
+ * </ul>
+ *
+ * <p> The tree structure supports confinement checks. The phrase "threads contained in
+ * the task scope" in method descriptions means threads started in the task scope or
+ * descendant scopes. StructuredTaskScope does not define APIs that exposes the tree
+ * structure at this time.
+ *
+ * <p> Unless otherwise specified, passing a {@code null} argument to a constructor
+ * or method in this class will cause a {@link NullPointerException} to be thrown.
+ *
+ * <h2>Memory consistency effects</h2>
+ *
+ * <p>Actions in the owner thread of, or a thread contained in, the task scope prior to
+ * {@linkplain #fork forking} of a {@code Callable} task
+ * <a href="../../../../java.base/java/util/concurrent/package-summary.html#MemoryVisibility">
+ * <i>happen-before</i></a> any actions taken by that task, which in turn <i>happen-before</i>
+ * the task result is retrieved via its {@code Future}, or <i>happen-before</i> any actions
+ * taken in a thread after {@linkplain #join() joining} of the task scope.
+ *
```

Comment:
Would a @-jls reference to the appropriate section of the memory model chapter help here?

Reply:
Yes, it can reference JLS 17.4.5.